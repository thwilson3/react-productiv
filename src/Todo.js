import React from "react";

/** Simple presentation component for a todo.
 *
 * Props:
 * - todo: like { id, title, description, priority }
 *
 * { EditableTodo, TopTodo } -> Todo
 **/

function Todo({todo}) {

  const {title, priority, description} = todo;
  return (
      <div id={`todo-${todo.id}`} className="Todo">
        <div><b>{title}</b> <small>(priority: {priority})</small></div>
        <div><small>{description}</small></div>
      </div>
  );
}

export default Todo;
