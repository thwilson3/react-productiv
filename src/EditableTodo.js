import React, { useState } from "react";
import Todo from "./Todo";
import TodoForm from "./TodoForm";

/** Show editable todo item.
 *
 * Props
 * - todo
 * - update(): fn to call to update a todo
 * - remove(): fn to call to remove a todo
 *
 * State
 * - isEditing: determines whether to show form or not
 *
 * EditableTodoList -> EditableTodo -> { Todo, TodoForm }
 */

function EditableTodo({ todo, update, remove }) {
  const [isEditing, setIsEditing] = useState(false);

  /** Toggle if this is being edited */
  function toggleEdit() {
    setIsEditing((e) => !e);
  }

  /** Call remove fn passed to this. */
  function handleDelete() {
    remove(todo.id);
  }

  /** Edit form saved; toggle isEditing and update in ancestor. */
  function handleSave(formData) {
    update(formData);
    toggleEdit();
  }

  return (
    <div className="EditableTodo">
      {isEditing
        ? <TodoForm initialFormData={todo} handleSave={handleSave} />
        : (
          <div className="mb-3">
            <div className="float-end text-sm-end">
              <button
                id={`edit-button-${todo.id}`}
                className="EditableTodo-toggle btn-link btn btn-sm"
                onClick={toggleEdit}
              >
                Edit
              </button>
              <button
                className="EditableTodo-delBtn btn-link btn btn-sm text-danger"
                onClick={handleDelete}
              >
                Del
              </button>
            </div>
            <Todo todo={todo} />
          </div>
        )}
    </div>
  );
}

export default EditableTodo;
