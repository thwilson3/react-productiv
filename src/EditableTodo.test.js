import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Todo from "./Todo";
import { initialTodos } from "./App";
import { update, remove } from "./TodoApp";
import EditableTodo from "./EditableTodo";

describe("EditableTodo", function () {
  it("renders without crashing", function () {
    render(
      <EditableTodo todo={initialTodos[0]} update={update} remove={remove} />
    );
  });

  it("renders edit button", function () {
    const result = render(
      <EditableTodo todo={initialTodos[0]} update={update} remove={remove} />
    );

    expect(
      result.container.querySelector(`#edit-button-${initialTodos[0].id}`)
    ).toBeInTheDocument();
  });

  it("shows form when edit button clicked", function () {
    const result = render(
      <EditableTodo todo={initialTodos[0]} update={update} remove={remove} />
    );

    fireEvent.click(
      result.container.querySelector(`#edit-button-${initialTodos[0].id}`)
    );

    expect(
      result.container.querySelector(`#todo-form-${initialTodos[0].id}`)
    ).toBeInTheDocument();

    expect(
      result.container.querySelector(`#edit-button-${initialTodos[0].id}`)
    ).not.toBeInTheDocument();
  });

  it("matches snapshot", function () {
    const { container } = render(
      <EditableTodo todo={initialTodos[0]} update={update} remove={remove} />
    );
    expect(container).toMatchSnapshot();
  });
});
