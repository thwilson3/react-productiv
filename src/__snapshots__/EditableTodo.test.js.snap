// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableTodo matches snapshot 1`] = `
<div>
  <div
    class="EditableTodo"
  >
    <div
      class="mb-3"
    >
      <div
        class="float-end text-sm-end"
      >
        <button
          class="EditableTodo-toggle btn-link btn btn-sm"
          id="edit-button-1"
        >
          Edit
        </button>
        <button
          class="EditableTodo-delBtn btn-link btn btn-sm text-danger"
        >
          Del
        </button>
      </div>
      <div
        class="Todo"
      >
        <div>
          <b>
            Code!
          </b>
           
          <small>
            (priority: 
            2
            )
          </small>
        </div>
        <div>
          <small>
            Write some code
          </small>
        </div>
      </div>
    </div>
  </div>
</div>
`;
